pkg/reconciler/managed/reconciler.go

```diff
@@ -725,26 +750,27 @@ func (r *Reconciler) Reconcile(_ context.Context, req reconcile.Request) (reconc
 			return reconcile.Result{Requeue: true}, errors.Wrap(r.client.Status().Update(ctx, managed), errUpdateManagedStatus)
 		}
 
-		if creation.ExternalNameAssigned {
-			en := meta.GetExternalName(managed)
-			// We will retry in all cases where the error comes from the api-server.
-			// At one point, context deadline will be exceeded and we'll get out
-			// of the loop. In that case, we warn the user that the external resource
-			// might be leaked.
-			err := retry.OnError(retry.DefaultRetry, resource.IsAPIError, func() error {
-				nn := types.NamespacedName{Name: managed.GetName()}
-				if err := r.client.Get(ctx, nn, managed); err != nil {
-					return err
-				}
-				meta.SetExternalName(managed, en)
-				return r.client.Update(ctx, managed)
-			})
-			if err != nil {
-				log.Debug("Cannot update managed resource", "error", err)
-				record.Event(managed, event.Warning(reasonCannotUpdateManaged, errors.Wrap(err, errUpdateManagedAfterCreate)))
-				managed.SetConditions(xpv1.ReconcileError(errors.Wrap(err, errUpdateManagedAfterCreate)))
-				return reconcile.Result{Requeue: true}, errors.Wrap(r.client.Status().Update(ctx, managed), errUpdateManagedStatus)
+		meta.SetExternalCreateTime(managed, metav1.Now())
+
+		// We try to persist annotation several times because if we fail
+		// to persist the external name annotation (if necessary) we'll
+		// leak our newly created external resource. Any other changes
+		// made during Create will be reverted; at the time of writing
+		// Create implementations are advised not to alter status, but
+		// we may revisit this in future.
+		a := managed.GetAnnotations()
+		if err := retry.OnError(retry.DefaultRetry, resource.IsAPIError, func() error {
```

Comment:
Note that we run this update with retries only in case external name assigned because we lose the status when we do this operation and there are cases where information is only available as result of creation output. One possible solution could be to save the creation time before we call `Create` though that would require two updates with retries, the one after create being inside conditional, which increases the complexity here.